x/posts/keeper/msg_server.go

```diff
@@ -498,3 +498,88 @@ func (m msgServer) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParam
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+// MovePost defines the rpc method for Msg/MovePost
+func (k msgServer) MovePost(goCtx context.Context, msg *types.MsgMovePost) (*types.MsgMovePostResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Make sure the post exists
+	post, found := k.GetPost(ctx, msg.SubspaceID, msg.PostID)
+	if !found {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "post with id %d does not exist", msg.PostID)
+	}
+
+	// Check if the target subspace exists
+	if !k.HasSubspace(ctx, msg.TargetSubspaceID) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.TargetSubspaceID)
+	}
+
+	// Check if the target section exists
+	if !k.HasSection(ctx, msg.TargetSubspaceID, msg.TargetSectionID) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace section with id %d not found", msg.TargetSectionID)
+	}
+
+	// Check the permission to move the post
+	canMove := post.Author == msg.Owner && k.HasPermission(ctx, msg.TargetSubspaceID, msg.TargetSectionID, msg.Owner, types.PermissionWrite)
+	if !canMove {
+		return nil, errors.Wrap(subspacestypes.ErrPermissionDenied, "you cannot move this post")
+	}
+
+	// Get the next post id
+	newPostID, err := k.GetNextPostID(ctx, msg.TargetSubspaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Update the post and validate it
+	updateTime := ctx.BlockTime()
+	postMove := types.NewPostMove(msg.TargetSubspaceID, msg.TargetSectionID, newPostID, updateTime)
+	updatedPost := postMove.Update(post)
```

Comment:
Why is the `PostMove` type defined? Why can't we simply use the `Post` type and update the fields?

Reply:
I refer the [`EditPost`](https://github.com/desmos-labs/desmos/blob/master/x/posts/keeper/msg_server.go#L151) update section to have update object to update fields since I think it brings the benefit as follows:

#### 1. msg server can be focus on operating store rather than updating post fields
If we simply updates the fields in this section, the codes will make the section longer and not only be focus on store operation as follows:
```go
...checking section

// Update the post and validate it
updateTime := ctx.BlockTime()
post.SubspaceID = msg.TargetSubspaceID
post.ID = newPostID
post.Conversation = 0
post.References = nil
// ...other upcoming new fields update

err = k.ValidatePost(ctx, updatedPost)
if err != nil {
return nil, err
}

k.SavePost(ctx, post)

...other store operations
```

#### 2. Having [specific unit test](https://github.com/desmos-labs/desmos/blob/paul/DCD-324/impl-post-transfer/x/posts/types/models_test.go#L1041) on `PostMove.Update` can have another protection for fields update

`TestPostMove` tests all the possibility on the fields update of the moved post, it can exclude the missing cases inside msg server test. As a result, we can be focus on the store operation inside msg server since the unit test guarantees fields update would be not wrong.